src/vdaf/prg.rs

```diff
@@ -232,6 +237,116 @@ impl SeedStream for SeedStreamSha3 {
     }
 }
 
+/// PrgFixedKeyAes128 as specified in [[draft-irtf-cfrg-vdaf-05]]. This PRG is NOT RECOMMENDED for
+/// general use; see Section 9 ("Security Considerations") for details.
+///
+/// This PRG combines SHA-3 and a fixed-key mode of operation for AES-128. The key is "fixed" in
+/// the sense that it is derived (using cSHAKE128) from the customization and binder strings, and
+/// depending on the application, these strings can be hard-coded. The seed is used construct each
+/// block of input passed to a hash function built from AES-128.
+///
+/// [draft-irtf-cfrg-vdaf-05]: https://datatracker.ietf.org/doc/draft-irtf-cfrg-vdaf/05/
+#[derive(Clone, Debug)]
+#[cfg(feature = "experimental")]
+pub struct PrgFixedKeyAes128 {
+    fixed_key_deriver: CShake128,
+    base_block: Block,
+}
+
+#[cfg(feature = "experimental")]
+impl Prg<16> for PrgFixedKeyAes128 {
+    type SeedStream = SeedStreamFixedKeyAes128;
+
+    fn init(seed_bytes: &[u8; 16], custom: &[u8]) -> Self {
+        Self {
+            fixed_key_deriver: CShake128::from_core(CShake128Core::new(custom)),
+            base_block: (*seed_bytes).into(),
+        }
+    }
+
+    fn update(&mut self, data: &[u8]) {
+        Update::update(&mut self.fixed_key_deriver, data);
+    }
+
+    fn into_seed_stream(self) -> SeedStreamFixedKeyAes128 {
+        let mut fixed_key = GenericArray::from([0; 16]);
+        XofReader::read(&mut self.fixed_key_deriver.finalize_xof(), &mut fixed_key);
+        SeedStreamFixedKeyAes128 {
+            base_block: self.base_block,
+            cipher: Aes128::new(&fixed_key),
+            length_consumed: 0,
+        }
+    }
+}
+
+/// Seed stream for [`PrgFixedKeyAes128`].
+#[cfg(feature = "experimental")]
+pub struct SeedStreamFixedKeyAes128 {
+    cipher: Aes128,
+    base_block: Block,
+    length_consumed: usize,
+}
+
+#[cfg(feature = "experimental")]
+impl SeedStream for SeedStreamFixedKeyAes128 {
+    fn fill(&mut self, buf: &mut [u8]) {
+        let next_length_consumed = self.length_consumed + buf.len();
+        let mut offset = self.length_consumed & 0b1111;
+        let mut index = 0;
+        let mut block = Block::from([0; 16]);
+
+        // NOTE(cjpatton) We might be able to speed this up by unrolling this loop and encrypting
+        // multiple blocks at the same time via `self.cipher.encrypt_blocks()`.
+        for block_counter in self.length_consumed >> 4..next_length_consumed >> 4 {
```

Comment:
This loop incorrectly does nothing if the stream pointer doesn't advance to the next block. The IDPF's `extend()` calls this with a length of 1, which hits this issue.

It occurs to me that we could use RustCrypto's traits for buffer management to implement this, but that sounds like an entire project unto itself.

Reply:
Fixed, and added a test for making sure a buffer gets filled even if it's less than a block.